### PR TITLE
chore: Add batch reference ADR

### PR DESCRIPTION
**Pull request by marikaner**
Closes SAP/cloud-sdk-backlog#813.
Relates to https://github.com/SAP/cloud-sdk-js/issues/2708.

### CODE REVIEW

File: knowledge-base/adr/0036-batch-references.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+```ts
+/* extend request type with getBatchRef() method */
+interface CreateRequestBuilder<EntityT> {
+  ...
+  getBatchRef(): BatchRef
+}
+
+interface BatchRef {
+  ref: string;
```
Comment (marikaner) [inline knowledge-base/adr/0036-batch-references.md:85]:
```suggestion
  id: string;
```

File: knowledge-base/adr/0036-batch-references.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+Currently the request ID (content-id of the batch request headers) is created on `execute()`. This would yield a new request ID on every request execution. The ID would neither be available nor reusable for sharing with other requests.
+In the future this should happen when instantiating a request. Users should be able to overwrite this request id.
+
+API (`setBatchRef()`):
+
+```ts
+interface CreateRequestBuilder<EntityT> {
+  ...
+  setBatchRef(ref: string): void;
```
Comment (marikaner) [inline knowledge-base/adr/0036-batch-references.md:156]:
```suggestion
  setBatchId(id: string): void;
```

File: knowledge-base/adr/0036-batch-references.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+Referencing batch becomes possible. The batch references are simple objects, which require wrapping for bound functions. The entity created from the entity builder could be confused with an actual entity.
+
+### Appendix
+
+#### Repeatability
+
+Currently the request ID (content-id of the batch request headers) is created on `execute()`. This would yield a new request ID on every request execution. The ID would neither be available nor reusable for sharing with other requests.
+In the future this should happen when instantiating a request. Users should be able to overwrite this request id.
+
+API (`setBatchRef()`):
```
Comment (marikaner) [inline knowledge-base/adr/0036-batch-references.md:151]:
```suggestion
API (`setBatchReference()`):
```

File: knowledge-base/adr/0036-batch-references.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+  .requestBuilder()
+  .create(address)
+  .asChildOf(
+    businessPartnerRef,
+    businessPartnerApi.schema.TO_BUSINESS_PARTNER_ADDRESS
+  );
+
+/* request 3 - bound function with reference to newly created business partner => let's assume there is a bound function in businessPartner */
+const functionRequest = businessPartnerApi.entityBuilder
+  .fromBatchRef(businessPartnerRef)
```
Comment (marikaner) [inline knowledge-base/adr/0036-batch-references.md:134]:
```suggestion
  .fromBatchReference(businessPartnerRef)
```

File: knowledge-base/adr/0036-batch-references.md [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+This is the same example as described in the current state, only with batch.
+
+```ts
+/* request 1 */
+const createRequest = businessPartnerApi
+  .requestBuilder()
+  .create(businessPartner);
+
+/* reference to newly created business partner */
+const businessPartnerRef = createRequest.getBatchRef();
```
Comment (marikaner) [inline knowledge-base/adr/0036-batch-references.md:121]:
```suggestion
const businessPartnerRef = createRequest.getBatchReference();
```

File: knowledge-base/adr/0036-batch-references.md [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+- createAsChildOf
+- delete
+- update
+- unbound functions/actions with entity parameters
+
+However, this will not work for bound functions, as they rely on the physical existence of an entity.
+
+To solve this we should have a new method on the entity builder, that allows building an entity based on a reference only.
+
+API (`fromBatchRef()`):
```
Comment (marikaner) [inline knowledge-base/adr/0036-batch-references.md:101]:
```suggestion
API (`fromBatchReference()`):
```

File: knowledge-base/adr/0036-batch-references.md [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+API (`getBatchRef()`):
+
+```ts
+/* extend request type with getBatchRef() method */
+interface CreateRequestBuilder<EntityT> {
+  ...
+  getBatchRef(): BatchRef
+}
+
+interface BatchRef {
```
Comment (marikaner) [inline knowledge-base/adr/0036-batch-references.md:84]:
```suggestion
interface BatchReference {
```

File: knowledge-base/adr/0036-batch-references.md [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+Delete and get all probably does not make sense for this.
+
+API (`getBatchRef()`):
+
+```ts
+/* extend request type with getBatchRef() method */
+interface CreateRequestBuilder<EntityT> {
+  ...
+  getBatchRef(): BatchRef
```
Comment (marikaner) [inline knowledge-base/adr/0036-batch-references.md:81]:
```suggestion
  getBatchReference(): BatchReference
```

File: knowledge-base/adr/0036-batch-references.md [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+Those requests are:
+
+- get by key
+- create
+- update
+- bound/unbound functions/actions
+
+Delete and get all probably does not make sense for this.
+
+API (`getBatchRef()`):
```
Comment (marikaner) [inline knowledge-base/adr/0036-batch-references.md:75]:
```suggestion
API (`getBatchReference()`):
```

File: knowledge-base/adr/0036-batch-references.md [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+    businessPartnerRef,
+    businessPartnerApi.schema.TO_BUSINESS_PARTNER_ADDRESS
+  );
+
+/* request 3 - bound function with reference to newly created business partner => let's assume there is a bound function in businessPartner */
+const functionRequest = businessPartnerApi.entityBuilder
+  .fromBatchReference(businessPartnerRef)
+  .doSomething();
+
+const batchResponse = await changeset(createRequest, asChildOfRequest, functionRequest).execute(...);
```
Comment (deekshas8) [inline knowledge-base/adr/0036-batch-references.md:137]: Should this be : await batch(changeset(createRequest, asChildOfRequest), functionRequest).execute(...); ?

Comment (marikaner) [reply knowledge-base/adr/0036-batch-references.md:137]: Yes, almost. Will fix it 😊

Comment (marikaner) [inline knowledge-base/adr/0036-batch-references.md:137]:
```suggestion
const batchResponse = await batch(
  changeset(createRequest, asChildOfRequest, functionRequest)
).execute(...);
```

File: knowledge-base/adr/0036-batch-references.md [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+Referencing batch becomes possible. The batch references are simple objects, which require wrapping for bound functions. The entity created from the entity builder could be confused with an actual entity.
+
+### Appendix
+
+#### Repeatability
+
+Currently the request ID (content-id of the batch request headers) is created on `execute()`. This would yield a new request ID on every request execution. The ID would neither be available nor reusable for sharing with other requests.
+In the future this should happen when instantiating a request. Users should be able to overwrite this request id.
+
+API (`setBatchReference()`):
```
[the remark this answers is not in the document]
Comment (marikaner) [reply knowledge-base/adr/0036-batch-references.md:151]: 👍 good catch

File: knowledge-base/adr/0036-batch-references.md [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+.requestBuilder()
+.create(address)
+.asChildOf(newBusinessPartner, businessPartnerApi.schema.TO_BUSINESS_PARTNER_ADDRESS);
+
+const newChildAddress = await asChildOfRequest.execute(...);
+
+/* request 3 - bound function with reference to newly created business partner => let's assume there is a bound function in businessPartner */
+const functionRequest = newBusinessPartner.getNumOfAddresses();
+
+const numOfAddresses = await functionRequest.execute(...);
```
Comment (marikaner) [inline knowledge-base/adr/0036-batch-references.md:60]:
```suggestion
const actionRequest = newBusinessPartner.getNumOfAddresses();

const numOfAddresses = await actionRequest.execute(...);
```

**Pull request merged**